### PR TITLE
Add Fedora base images

### PR DESCRIPTION
**Pull request by nghiant2710**
Since Fedora base images are very experimental, we will only support artik5 and artik10. Other boards will be added later.

### CODE REVIEW

**Comment by petrosagg** [comment]
let's have only one autogenerated commit per PR

**Comment by nghiant2710** [comment]
@petrosagg, PR updated! Will have only one commit for autogenerated Dockerfiles from the next PR

File: systemd/Dockerfile.fedora.tpl [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+RUN rpm --rebuilddb && yum install -y \
+        ca-certificates tar \
+        systemd \
+    && yum clean all
+
+# Install Systemd
+
+ENV container docker
+RUN (cd /lib/systemd/system/sysinit.target.wants/; for i in *; do [ $i == systemd-tmpfiles-setup.service ] || rm -f $i; done); \
```
Comment (petrosagg) [inline systemd/Dockerfile.fedora.tpl]: Why do we delete all these files? We should touch /lib folders, the correct way to do it is via symlinks in /etc. `systemd disable` puts them there automatically

File: systemd/Dockerfile.fedora.tpl [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        systemd-logind.service \
+        systemd-remount-fs.service \
+        getty.target \
+        graphical.target \
+        tmp.mount \
+        console-getty.service \
+        auditd \
+    && rm -f /etc/systemd/system/default.target \
+    && ln -s /lib/systemd/system/multi-user.target /etc/systemd/system/default.target \
+    && cat /dev/null > /etc/fstab
```
Comment (petrosagg) [inline systemd/Dockerfile.fedora.tpl]: You can use `truncate /etc/fstab`, but why are we emptying this file?

File: buildpack-deps/Dockerfile.fedora.tpl [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+		libjpeg-devel \
+		ImageMagick-devel \
+		mysql-devel \
+		ncurses-devel \
+		libpq-devel \
+		libreadline-devel \
+		sqlite-devel \
+		libsqlite3 \
+		openssl \
+    	openssl-devel \
```
Comment (petrosagg) [inline buildpack-deps/Dockerfile.fedora.tpl]: Indentation

File: python/Dockerfile.fedora.python3.slim.tpl [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+RUN set -x \
+	&& buildDeps=' \
+		curl \
+	' \
+	&& rpm --rebuilddb && yum install -y $buildDeps && yum clean all \
+	&& curl -SLO ftp://195.220.108.108/linux/opensuse/ports/update/13.1/armv7hl/libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm \
+	&& echo "73ad776d7a6cc5d2559b209499ae4b82d0ce329d2f9e2ce345d249350c34bf45  libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm" | sha256sum -c - \
+	&& rpm -ivh libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm \
+	&& rm libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.python3.slim.tpl]: Why are we compiling libopenssl from source?

Comment (petrosagg) [inline python/Dockerfile.fedora.python3.slim.tpl]: Ok, not compiling. But why are we using a package from opensuse?

File: python/Dockerfile.fedora.python3.slim.tpl [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	&& curl -SLO "#{BINARY_URL}" \
+	&& echo "#{CHECKSUM}" | sha256sum -c - \
+	&& tar -xzf "Python-$PYTHON_VERSION.linux-#{TARGET_ARCH}.tar.gz" --strip-components=1 \
+	&& rm -rf "Python-$PYTHON_VERSION.linux-#{TARGET_ARCH}.tar.gz" \
+	&& echo "/usr/local/lib" >> /etc/ld.so.conf \
+	&& ldconfig \
+	&& mkdir -p /usr/src/python/setuptools \
+	&& curl -SLO https://github.com/pypa/setuptools/archive/v$SETUPTOOLS_VERSION.tar.gz \
+	&& echo "$SETUPTOOLS_SHA256  v$SETUPTOOLS_VERSION.tar.gz" > setuptools-$SETUPTOOLS_VERSION.tar.gz.sha256sum \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.python3.slim.tpl]: You can pipe directly to `sha256sum -c`

File: python/Dockerfile.fedora.python3.slim.tpl [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	&& sha256sum -c setuptools-$SETUPTOOLS_VERSION.tar.gz.sha256sum \
+	&& tar -xzC /usr/src/python/setuptools --strip-components=1 -f v$SETUPTOOLS_VERSION.tar.gz \
+	&& rm -rf v$SETUPTOOLS_VERSION.tar.gz* \
+	&& cd /usr/src/python/setuptools \
+	&& python3 setup.py install \
+	&& mkdir -p /usr/src/python/pip \
+	&& curl -SL "https://github.com/pypa/pip/archive/$PYTHON_PIP_VERSION.tar.gz" -o pip.tar.gz \
+	&& echo "$PYTHON_PIP_SHA256  pip.tar.gz" > pip.tar.gz.sha256sum \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.python3.slim.tpl]: Same here

File: python/Dockerfile.fedora.python3.slim.tpl [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	&& sha256sum -c pip.tar.gz.sha256sum \
+	&& tar -xzC /usr/src/python/pip --strip-components=1 -f pip.tar.gz \
+	&& rm pip.tar.gz* \
+	&& cd /usr/src/python/pip \
+	&& python3 setup.py install \
+	&& cd .. \
+	&& find /usr/local \
+		\( -type d -a -name test -o -name tests \) \
+		-o \( -type f -a -name '*.pyc' -o -name '*.pyo' \) \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.python3.slim.tpl]: Why are we removing those files?

File: python/Dockerfile.fedora.python3.tpl [hunk 8]
```diff
@@ -0,0 +1,71 @@
+FROM #{FROM}
+
+# http://bugs.python.org/issue19846
+# > At the moment, setting "LANG=C" on a Linux system *fundamentally breaks Python 3*, and that's not OK.
+ENV LANG C.UTF-8
+
+# install libopenssl v1.0.0
+RUN curl -SLO ftp://195.220.108.108/linux/opensuse/ports/update/13.1/armv7hl/libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm \
+	&& echo "73ad776d7a6cc5d2559b209499ae4b82d0ce329d2f9e2ce345d249350c34bf45  libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm" | sha256sum -c - \
+	&& rpm -ivh libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.python3.tpl]: Same

File: python/Dockerfile.fedora.python3.tpl [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+RUN set -x \
+	&& curl -SLO "#{BINARY_URL}" \
+	&& echo "#{CHECKSUM}" | sha256sum -c - \
+	&& tar -xzf "Python-$PYTHON_VERSION.linux-#{TARGET_ARCH}.tar.gz" --strip-components=1 \
+	&& rm -rf "Python-$PYTHON_VERSION.linux-#{TARGET_ARCH}.tar.gz" \
+	&& echo "/usr/local/lib" >> /etc/ld.so.conf \
+	&& ldconfig \
+	&& mkdir -p /usr/src/python/setuptools \
+	&& curl -SLO https://github.com/pypa/setuptools/archive/v$SETUPTOOLS_VERSION.tar.gz \
+	&& echo "$SETUPTOOLS_SHA256  v$SETUPTOOLS_VERSION.tar.gz" > setuptools-$SETUPTOOLS_VERSION.tar.gz.sha256sum \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.python3.tpl]: same

File: python/Dockerfile.fedora.python3.tpl [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	&& sha256sum -c setuptools-$SETUPTOOLS_VERSION.tar.gz.sha256sum \
+	&& tar -xzC /usr/src/python/setuptools --strip-components=1 -f v$SETUPTOOLS_VERSION.tar.gz \
+	&& rm -rf v$SETUPTOOLS_VERSION.tar.gz* \
+	&& cd /usr/src/python/setuptools \
+	&& python3 setup.py install \
+	&& mkdir -p /usr/src/python/pip \
+	&& curl -SL "https://github.com/pypa/pip/archive/$PYTHON_PIP_VERSION.tar.gz" -o pip.tar.gz \
+	&& echo "$PYTHON_PIP_SHA256  pip.tar.gz" > pip.tar.gz.sha256sum \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.python3.tpl]: same

File: python/Dockerfile.fedora.python3.tpl [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+	&& sha256sum -c pip.tar.gz.sha256sum \
+	&& tar -xzC /usr/src/python/pip --strip-components=1 -f pip.tar.gz \
+	&& rm pip.tar.gz* \
+	&& cd /usr/src/python/pip \
+	&& python3 setup.py install \
+	&& cd .. \
+	&& find /usr/local \
+		\( -type d -a -name test -o -name tests \) \
+		-o \( -type f -a -name '*.pyc' -o -name '*.pyo' \) \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.python3.tpl]: same

File: python/Dockerfile.fedora.slim.tpl [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+ENV PYTHON_PIP_SHA256 8dae1fb72e29c2b6ff6ed267861179216bf98d3bda6d30e527dbed0db5ac7e1d
+
+ENV SETUPTOOLS_SHA256 3f24380f0b18f869b7ad92d9e16d1d58093a9bac13bbd0b0e1fc5a82b460311f
+ENV SETUPTOOLS_VERSION 21.0.0
+
+RUN set -x \
+	&& buildDeps=' \
+		curl \
+	' \
+	&& rpm --rebuilddb && yum install -y $buildDeps && yum clean all \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.slim.tpl]: Shouldn't `yum clean all` be at the end of this `RUN`?

File: python/Dockerfile.fedora.slim.tpl [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+	&& curl -SLO ftp://195.220.108.108/linux/opensuse/ports/update/13.1/armv7hl/libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm \
+	&& echo "73ad776d7a6cc5d2559b209499ae4b82d0ce329d2f9e2ce345d249350c34bf45  libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm" | sha256sum -c - \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.slim.tpl]: same

File: python/Dockerfile.fedora.slim.tpl [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+	&& rm libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm \
+	&& curl -SLO "#{BINARY_URL}" \
+	&& echo "#{CHECKSUM}" | sha256sum -c - \
+	&& tar -xzf "Python-$PYTHON_VERSION.linux-#{TARGET_ARCH}.tar.gz" --strip-components=1 \
+	&& rm -rf "Python-$PYTHON_VERSION.linux-#{TARGET_ARCH}.tar.gz" \
+	&& echo "/usr/local/lib" >> /etc/ld.so.conf \
+	&& ldconfig \
+	&& mkdir -p /usr/src/python/setuptools \
+	&& curl -SLO https://github.com/pypa/setuptools/archive/v$SETUPTOOLS_VERSION.tar.gz \
+	&& echo "$SETUPTOOLS_SHA256  v$SETUPTOOLS_VERSION.tar.gz" > setuptools-$SETUPTOOLS_VERSION.tar.gz.sha256sum \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.slim.tpl]: same

File: python/Dockerfile.fedora.slim.tpl [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+	&& sha256sum -c setuptools-$SETUPTOOLS_VERSION.tar.gz.sha256sum \
+	&& tar -xzC /usr/src/python/setuptools --strip-components=1 -f v$SETUPTOOLS_VERSION.tar.gz \
+	&& rm -rf v$SETUPTOOLS_VERSION.tar.gz* \
+	&& cd /usr/src/python/setuptools \
+	&& python2 setup.py install \
+	&& mkdir -p /usr/src/python/pip \
+	&& curl -SL "https://github.com/pypa/pip/archive/$PYTHON_PIP_VERSION.tar.gz" -o pip.tar.gz \
+	&& echo "$PYTHON_PIP_SHA256  pip.tar.gz" > pip.tar.gz.sha256sum \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.slim.tpl]: same

File: python/Dockerfile.fedora.slim.tpl [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+	&& sha256sum -c pip.tar.gz.sha256sum \
+	&& tar -xzC /usr/src/python/pip --strip-components=1 -f pip.tar.gz \
+	&& rm pip.tar.gz* \
+	&& cd /usr/src/python/pip \
+	&& python2 setup.py install \
+	&& cd .. \
+	&& find /usr/local \
+		\( -type d -a -name test -o -name tests \) \
+		-o \( -type f -a -name '*.pyc' -o -name '*.pyo' \) \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.slim.tpl]: same

File: python/Dockerfile.fedora.tpl [hunk 17]
```diff
@@ -0,0 +1,62 @@
+FROM #{FROM}
```
Comment (petrosagg) [inline python/Dockerfile.fedora.tpl]: same comments for this file too

File: systemd/Dockerfile.fedora.tpl (hunk 1 above)
Comment (nghiant2710) [inline systemd/Dockerfile.fedora.tpl]: I did `systemd disable` lot of services but still got tons of noisy logs when systemd start then I found this article about systemd on fedora container http://developerblog.redhat.com/2014/05/05/running-systemd-within-docker-container/. Basically, this is a fast and clean way to stop systemd from stating many services (we only keep  systemd, journald, udev and resin services running), it's way better than disable every service since this image is not optimised for Docker container.
Until they release the armhf fedora Docker images, I think we should follow this way

File: systemd/Dockerfile.fedora.tpl (hunk 2 above)
Comment (nghiant2710) [inline systemd/Dockerfile.fedora.tpl]: it comes from this bug https://fedorahosted.org/spin-kickstarts/ticket/54

File: python/Dockerfile.fedora.python3.slim.tpl (hunk 4 above)
Comment (nghiant2710) [inline python/Dockerfile.fedora.python3.slim.tpl]: python `ssl` lib requires libopenssl v1.0.0 however, that version is not available on fedora (only v1.0.1) so I use this package and it works tho.

File: python/Dockerfile.fedora.python3.slim.tpl (hunk 5 above)
Comment (nghiant2710) [inline python/Dockerfile.fedora.python3.slim.tpl]: Will update this

File: python/Dockerfile.fedora.python3.slim.tpl (hunk 7 above)
Comment (nghiant2710) [inline python/Dockerfile.fedora.python3.slim.tpl]: those are compiled python scripts and can be removed after finishing building setuptools and pip

**Comment by nghiant2710** [comment]
@petrosagg these base images are built on top of fedora minimal armhf images which are not optimised for Docker container.
I'm following their armhf docker images here https://ausil.us/fedora-arm/, as soon as they release them or the issue with dnf is resolved, we will switch to these images

**Comment by nghiant2710** [comment]
@petrosagg I've rebased this PR and addressed all your comments

File: golang/Dockerfile.fedora.slim.tpl [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+	&& curl -SLO "#{BINARY_URL}" \
+	&& echo "#{CHECKSUM}" | sha256sum -c - \
+	&& tar -xzf "go$GO_VERSION.linux-#{TARGET_ARCH}.tar.gz" -C /usr/local/go --strip-components=1 \
+	&& rm go$GO_VERSION.linux-#{TARGET_ARCH}.tar.gz
+
+ENV GOROOT /usr/local/go
+ENV GOPATH /go
+ENV PATH $GOPATH/bin:/usr/local/go/bin:$PATH
+
+RUN mkdir -p "$GOPATH/src" "$GOPATH/bin" && chmod -R 777 "$GOPATH"
```
Comment (petrosagg) [inline golang/Dockerfile.fedora.slim.tpl]: Can you move this RUN with the installation further up. It will make for one less layer

Comment (petrosagg) [inline golang/Dockerfile.fedora.slim.tpl]: Why the chmod?

File: python/Dockerfile.fedora.python3.slim.tpl [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+ENV SETUPTOOLS_SHA256 3f24380f0b18f869b7ad92d9e16d1d58093a9bac13bbd0b0e1fc5a82b460311f
+ENV SETUPTOOLS_VERSION 21.0.0
+
+RUN set -x \
+	&& buildDeps=' \
+		curl \
+	' \
+	&& rpm --rebuilddb && yum install -y $buildDeps \
+	&& curl -SLO ftp://195.220.108.108/linux/opensuse/ports/update/13.1/armv7hl/libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm \
+	&& echo "73ad776d7a6cc5d2559b209499ae4b82d0ce329d2f9e2ce345d249350c34bf45  libopenssl1_0_0-1.0.1e-11.9.1.armv7hl.rpm" | sha256sum -c - \
```
Comment (petrosagg) [inline python/Dockerfile.fedora.python3.slim.tpl]: I don't think this is a good idea at all. Fedora ships python3 packages so surely there is a way to compile python from source while linking to the right libraries

File: golang/Dockerfile.fedora.slim.tpl (hunk 18 above)
Comment (nghiant2710) [inline golang/Dockerfile.fedora.slim.tpl]: Updated to make one layer only.  Chmod is from official Dockerfile, however I think chmod will make sure `go` commands (e.g. go get or go install) will have no issue with permissions to write on $GOPATH

**Comment by nghiant2710** [comment]
@petrosagg, Python base images will temporarily use python packages from Fedora (they're well maintained and update to date with latest Python 2.x and 3.x). I'm working on building Python images with in-house package but having issue with broken dnf when using in-house python 3 binary. I will try to resolve it asap. However, we should merge this PR first then update Python base images later

File: device-base/generate-dockerfile.sh [hunk 20]
```diff
@@ -127,12 +128,16 @@ for device in $devices; do
...
+		fedora_template='Dockerfile.fedora.tpl'
+		fedora_baseImage='armhf-fedora'
 	;;
 	'artik10')
 		template='Dockerfile.tpl'
 		baseImage='armv7hf-debian'
 		alpine_template='Dockerfile.alpine.tpl'
 		alpine_baseImage='armhf-alpine'
+		fedora_template='Dockerfile.fedora.tpl'
+		fedora_baseImage='armhf-fedora'
```
Comment (petrosagg) [inline device-base/generate-dockerfile.sh]: For the other ARM images we use armv7hf, why the change?

File: buildpack-deps/Dockerfile.fedora.tpl [hunk 21]
```diff
@@ -0,0 +1,26 @@
+FROM #{FROM}
+
+RUN dnf install -y \
+		autoconf \
+		make automake gcc gcc-c++ kernel-devel \
```
Comment (petrosagg) [inline buildpack-deps/Dockerfile.fedora.tpl]: can you split these too and also sort them alphabetically?

File: golang/Dockerfile.fedora.slim.tpl [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+RUN buildDeps='gcc gcc-c++ git' \
+	&& set -x \
+	&& dnf install -y $buildDeps \
+	&& dnf clean all \
+	&& mkdir -p /usr/local/go \
+	&& curl -SLO "#{BINARY_URL}" \
+	&& echo "#{CHECKSUM}" | sha256sum -c - \
+	&& tar -xzf "go$GO_VERSION.linux-#{TARGET_ARCH}.tar.gz" -C /usr/local/go --strip-components=1 \
+	&& rm go$GO_VERSION.linux-#{TARGET_ARCH}.tar.gz \
+	&& mkdir -p "$GOPATH/src" "$GOPATH/bin" && chmod -R 777 "$GOPATH"
```
Comment (petrosagg) [inline golang/Dockerfile.fedora.slim.tpl]: Can you put `chmod` on its own line?

**Comment by petrosagg** [comment]
@nghiant2710 how are we going to correctly tag the python images if we're using python from Fedora's packages?

File: device-base/generate-dockerfile.sh (hunk 20 above)
Comment (nghiant2710) [inline device-base/generate-dockerfile.sh]: I expect this base image will work on armv6hf device (the same as alpine base image) but just test on a RPI model B+ and it doesn't work. So I will update this to armv7hf

File: buildpack-deps/Dockerfile.fedora.tpl (hunk 21 above)
Comment (nghiant2710) [inline buildpack-deps/Dockerfile.fedora.tpl]: there is no build-essential package on Fedora and this combination is the closest equivalent so I want to have them in one line

**Comment by nghiant2710** [comment]
@petrosagg, we won't have the full version tags for Python base image for now, only the major version e.g. `resin/artik5-fedora-python:2` or `resin/artik5-fedora-python:3`

**Comment by petrosagg** [comment]
ok. There are still pending comments. Please don't rebase while you resolve them. You can rebase/squash when you get the SGTM

File: buildpack-deps/Dockerfile.fedora.tpl (hunk 21 above)
Comment (petrosagg) [inline buildpack-deps/Dockerfile.fedora.tpl]: I see that there is a package named `@development-tools`, but in any case this package differs from distro to distro and doesn't necessarily contain the same build packages across all of them. Let's keep a consistent coding style

**Comment by petrosagg** [comment]
SGTM

File: buildpack-deps/Dockerfile.fedora.tpl (hunk 21 above)
Comment (nghiant2710) [inline buildpack-deps/Dockerfile.fedora.tpl]: @petrosagg `@development-tools` will install too many packages that won't be necessary in our base images so I don't want to use it.
I will sort the package name, rebase then merge this PR

Comment (petrosagg) [inline buildpack-deps/Dockerfile.fedora.tpl]: :+1:
